pkg/resource/deploy/step_generator.go

```diff
@@ -619,6 +614,122 @@ func (sg *stepGenerator) issueCheckErrors(new *resource.State, urn resource.URN,
+	if !ok {
+		return nil, errors.Errorf("unknown provider '%v' for resource '%v'", ref, urn)
+	}
+	return p, nil
+}
+
+func (sg *stepGenerator) calculateDependentReplacements(root *resource.State) ([]*resource.State, error) {
+	// We need to compute the set of resources that may be replaced by a change to the resource under consideration.
+	// We do this by taking the complete set of transitive dependents on the resource under consideration and
+	// removing any resources that would not be replaced by changes to their dependencies. We determine whether or not
+	// a resource may be replaced by substituting unknowns for input properties that may change due to deletion of the
+	// resources their value depends on and calling the resource provider's `Diff` method.
+	//
+	// This is perhaps clearer when described by example. Consider the following dependency graph:
+	//
+	//       A
+	//     __|__
+	//     B   C
+	//     |  _|_
+	//     D  E F
+	//
+	// In this graph, all of B, C, D, E, and F transitively depend on A. It may be the case, however, that changes to
+	// the specific properties of any of those resources R that would occur if a resource on the path to A were deleted
+	// and recreated may not cause R to be replaced. For example, the edge from B to A may be a simple `dependsOn` edge
+	// such that a change to B does not actually influence any of B's input properties. In that case, neither B nor D
+	// would need to be deleted before A could be deleted.
+	var toReplace []*resource.State
+	replaceSet := map[resource.URN]bool{root.URN: true}
+
+	requiresReplacement := func(r *resource.State) (bool, error) {
+		// Neither custom nor external resources require replacement.
+		if !r.Custom || r.External {
+			return false, nil
+		}
+
+		// If the resource's provider is in the delete set, we mustreplace this resource.
+		if r.Provider != "" {
+			ref, err := providers.ParseReference(r.Provider)
+			if err != nil {
+				return false, err
+			}
+			if replaceSet[ref.URN()] {
+				return true, nil
+			}
+		}
+
+		// Scan the properties of this resource in order to determine whether or not any of them depend on a resource
+		// that requires replacement and build a set of input properties for the provider diff.
+		hasDependencyInReplaceSet, inputsForDiff := false, resource.PropertyMap{}
+		for pk, pv := range r.Inputs {
+			for _, propertyDep := range r.PropertyDependencies[pk] {
+				if replaceSet[propertyDep] {
+					hasDependencyInReplaceSet = true
+					pv = resource.MakeComputed(resource.NewStringProperty("<unknown>"))
+				}
+			}
+			inputsForDiff[pk] = pv
+		}
+
+		// If none of this resource's properties depend on a resource in the replace set, then none of the properties
+		// may change and this resource does not need to be replaced.
+		if !hasDependencyInReplaceSet {
+			return false, nil
+		}
+
+		// Otherwise, fetch the resource's provider. Since we have filtered out component resources, this resource must
+		// have a provider.
```

Comment:
I'm having trouble convincing myself that this is correct in the presence of provider resources in the set of dependencies.

We can only load a provider if we've called `Check` on it, right? Is it possible that we can get to a resource in our dependency graph whose provider hasn't been visited yet? This is super subtle and it would be awesome to have some additional comments when walking the dependency graph and what order we do it.

Reply:
> We can only load a provider if we've called Check on it, right? Is it possible that we can get to a resource in our dependency graph whose provider hasn't been visited yet? This is super subtle and it would be awesome to have some additional comments when walking the dependency graph and what order we do it.

This is not possible, no. The dependency graph always uses old resources, and all old providers are loaded when the plan is created.